src/gss_names.c

```diff
@@ -626,7 +696,106 @@ uint32_t gssntlm_inquire_name(uint32_t *minor_status,
                               gss_OID *MN_mech,
                               gss_buffer_set_t *attrs)
 {
-    return GSS_S_UNAVAILABLE;
+    uint32_t retmin;
+    uint32_t retmaj;
+    uint32_t tmpmin;
+    const struct gssntlm_name *in = (const struct gssntlm_name *)name;
+
+    if (!attrs) {
+        return GSSERRS(ERR_NOARG, GSS_S_CALL_INACCESSIBLE_WRITE);
+    }
+    *attrs = GSS_C_NO_BUFFER_SET;
+
+    if (name == GSS_C_NO_NAME) {
+        return GSSERRS(GSS_S_BAD_NAME, GSS_S_CALL_INACCESSIBLE_READ);
+    }
+
+    for (size_t i = 0; in->attrs && in->attrs[i].attr_name != NULL; i++) {
+        struct gssntlm_name_attribute *attr = &in->attrs[i];
+        size_t attr_name_len = strlen(attr->attr_name);
+        gss_buffer_desc buf;
+        gss_buffer_t attr_value = &attr->attr_value;
+        /* +1 for '=' separator and +1 for EOL */
+        size_t full_string_len = attr_value->length + attr_name_len + 2;
+        size_t offset = 0;
+        char *attr_string = malloc(full_string_len);
+        if (attr_string == NULL) {
+            set_GSSERR(ENOMEM);
+            goto done;
+        }
+
+        /* Construct 'attr_name=<attr_value>\0' string */
+        memcpy(attr_string, attr->attr_name, attr_name_len);
+        offset += attr_name_len;
+
+        attr_string[offset++] = '=';
+
+        memcpy(attr_string + offset, attr_value->value, attr_value->length);
+        offset += attr_value->length;
+
+        attr_string[offset] = 0;
+
+        /* now add a buffer to output set */
+        buf.length = full_string_len;
+        buf.value = attr_string;
+        retmaj = gss_add_buffer_set_member(&retmin, &buf, attrs);
+        free(buf.value);
+        if (retmaj != GSS_S_COMPLETE) goto done;
+    }
+    return GSSERRS(0, GSS_S_COMPLETE);
+
+done:
```

Comment:
Sorry for not catching this earlier, but when we use the done: label we tend to mean the function only exits there (with the notable exception of some early returns on input verification).

if you initialize retmin and retmaj to 0 in the variable declaration we can easily change this code to drop the test and goto at the end of the for loop, drop the return right after the for loop and change the done: part to:

```
done:
    /* for loop here */
    }
    if (retmaj) {
        (void)gss_release_buffer_set(&tmpmin, attrs);
    }
    return GSSERR();
}
```

Reply:
yep, reasonable approach...
But the problem here is that if gss_add_buffer_set_member() failed,
it just sets retmaj/retmin and doesn't write anything to debug log. And GSSERR() macro just returns quietly, so no debug messages will be seen at all.
How do you cope with that?

In the similar place (see gssntlm_sspi_session_key) you use 'return GSSERRS(retmin, retmaj)' instead of 'return GSSERR()' BUT in this case error from 'set_GSSERR' will be reported twice (once from set_GSSERR and once from GSSERRS). In my function I have both cases - set_GSSERR and gss_add_buffer_set_member.
Is it better to return an error twice than not report it at all?

Reply:
Uhm I really haven't thought hard about that case, but I would say that it is a case by case basis.
Here for example the only case where gss_add_buffer_set_member(0 can fail is basically an out of memory error for now. So not logging specifically about it is probably ok. If you are out of memory it is generally because of multiple factors and knowing exactly that it happened in a specific function rather than another is rarely going to be interesting, as that is just the trigger, not the cause.
In other places it may be better to risk double debugging than nothing, but I will gladly accept patches to fix the double debugging as a separate PR.